source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -399,6 +406,23 @@ ReadOrParseState Filter::readProxyHeader(Network::ListenerFilterBuffer& buffer)
   auto raw_slice = buffer.rawSlice();
   const char* buf = static_cast<const char*>(raw_slice.mem_);
 
+  if (config_.get()->allowRequestsWithoutProxyProtocol()) {
+    if (raw_slice.len_ < PROXY_PROTO_V1_SIGNATURE_LEN) {
+      ENVOY_LOG(debug, "request does not have enough bytes to determine if v1 or v2 proxy "
+                       "protocol, forwarding as is");
+      return ReadOrParseState::SkipFilter;
```

Comment:
I considered returning `ReadOrParseState::TryAgainLater` here to wait for more bytes to make the right determination, but I don't think we can tell the difference between a partial proxy protocol request and a non-proxy protocol request that just has fewer bytes than the proxy protocol headers.

i.e. waiting for more bytes (that may never come, e.g. here) means this test will infinite loop:
```c++
TEST_P(ProxyProtocolTest, AllowTinyNoProxyProtocol) {
  // allows a small request (less bytes than v1/v2 header) through even though it doesn't use proxy
  // protocol
  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
  proto_config.set_allow_requests_without_proxy_protocol(true);
  connect(true, &proto_config);

  std::string msg = "data";
  EXPECT_GT(PROXY_PROTO_V1_SIGNATURE_LEN, msg.length());
  EXPECT_GT(PROXY_PROTO_V2_HEADER_LEN, msg.length());

  write(msg);

  expectData(msg);

  disconnect();
}
```

instead I chose to treat any request with fewer bytes (I perceive this as rare, but perhaps not on bad networks? I would love feedback here) as NOT proxy protocol, and pass it through unchanged.

The open question is 'how likely is the healthcheck request is similar to proxy proxy signature?' If we decide this is rare, we could also just compare what we have and use that -- the risk there being e.g. a one byte POST/PUT request matching the `P` that also starts proxy protocol.

Reply:
I was curious how this logic handles the case of invalid proxy protocol v2 being sent, does this also just wait until timeout? https://github.com/envoyproxy/envoy/blob/3dc4ebf32e38099aff55544bfe74daa7d62be147/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L448-L449

If this is sane, I might be able to copy the approach but upon first glance this looked like a potential bug to me (although the user would have to configure proxy protocol v2 and then send most of but not all of a valid proxy protocol v2)

Reply:
I think you only can wait for enough bytes for proxy v1 and v2 header, otherwise, you can't confirm it is slow or just a partial header. For the partial header, you only can wait for a timeout.

Reply:
> I think you only can wait for enough bytes for proxy v1 and v2 header, otherwise, you can't confirm it is slow or just a partial header. For the partial header, you only can wait for a timeout.

That means you have two places to know the header is invalid, one for v2
https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abL409

another one is for v1
https://github.com/envoyproxy/envoy/pull/18951/files#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abL460

Reply:
I think you wait for enough bytes, and document that with this filter, requests of less than X bytes will always timeout.